### ISSUE TITLE
Apps with permissions to multiple folders

### ISSUE THREAD

**Issue by jonaswre** [Peergos/Peergos#970]
I have a use-case where I want to access multiple folders. An example app would be a gallery that can view your own photos directory and also photo directories that were shared with you. 

The main peergos app would need to shows a Permission Dialog where you can explore the file tree and add files that can be accessed by the app. 

Same could be done for read and write of folders/files.

And there is another related thing that would be cool. If an app can share a sub folder of the app storage with one or more friends or maybe even with your followers. And then your friends and followers can access those subfolders through their own install of the app. This could also be very helpful for a gallery as well. Sharing content from a context related ui without the UX break when sharing files from the normal file explorer.

I'm not sure but I think if you let an app share it's internal data, the app on the recipient side wouldn't need to ask for permissions to read the shared state. Since that state is not your personal information and the sender explicitly wanted to share it.

**Comment by jonaswre** [comment]
This might be dangerous?! 

I was just in general think about the actual usefulness of read permissions. If the app can't exfiltrate information why do you need to limit the read permission? It can't manipulate the data and it can't steal it what harm can a rouge app cause? 

But write and share are dangerous in combination. There might need to be a mechanism to make this understood to the user. Otherwise you will still need to trust the app author. 

If an app has read and write permissions every write to a shared location needs to be allowed. If an app has no read permissions there is no harm in writing content in some shared location. And if both use cases don't work a application needs to be able to flag as trusted that needs to be very explicit.

**Comment by ianopolous** [comment]
Defense in depth is the real long-term answer. But technically, right now, exfiltration is possible until browsers implement the CSP to block DNS prefetch and webrtc (there are links in the tech book apps page). 

Another permission we want to think through and add is along the lines of read access to the news feed (maybe restricted by mimetype). This would allow people to implement their own newsfeed viewers with different algorithms/persistence.

**Comment by jonaswre** [comment]
I was thinking about attack Vektors that come from peergos. The browser sandboxing is nothing we can do about for now. 

Newsfeed permissions would be cool but I think we still need a way for an app to request to share parts of it's state. Or maybe have an app specific Newsfeed. I'll have to get a better understanding how the feed works. 

I'll need to get some of my friends on peergos as well! 

But the most important permission that's missing is to request read access for multiple folders.

**Comment by ianopolous** [comment]
Agreed on an app requesting to share some of its state. This probably doesn't need a permission because the user has to decide who to share with anyway, just an API. 

Signups on peergos.net are currently open, but if you need more invites later then of course just email or message me. 

I'm still trying to understand the use case for multiple folder read access. It sounds like what you really want for the gallery app is access to all files of a certain type (images).  If there was a permission to read all files of a type, and an API to enumerate their paths would that work? This would also work for a music player where all my music was spread out over multiple folders in different subtrees.

**Comment by jonaswre** [comment]
Okay so I think this depends on whats supported when sharing. Can I share files with followers, friends or also just by username? 

If I can share files with all users I need permissions otherwise an app can leak data to a user. Even if I can only share with friends I would be worried about an app just sharing all my files with all my friends. Even if I just want my app to be able to share with specific people. Obviously an app could implement sharing with specific people but that should be enforced by peergos not up to the app to honor this.

So for the gallery app being able to read all image/jpeg wouldn't work. Because I might have images that are not relevant for the gallery (Game assets, Projects files,...) I mean if I allow access to all images I'm not sure  why we need read permissions at all? - The only reason I can think of is to mitigate if either trust or isolation fails. But then I would want to isolate an app to specific location not file types. Because the location usually cotains information in the path. A folder "Photos" might have images and video but only things that were take by me with a camera. Where as "Documents" might also have images but contain a very different kind of data that maybe should be isolated from the gallery app.

Docker also isolates access by path not mime type just as a side note

**Comment by ianopolous** [comment]
You can't share with someone who isn't following you. But you can share with any number of followers by name, all followers or friends, or hopefully soon custom groups that you define (chosen from your followers). 

The sharing API would require the user to okay each share (the same as current sharing), which is why I say it might not need a permission. 

Okay I think I'm convinced on the need for a permission for read access to multiple paths. :-) This is the first instance of a permission that needs parameters. What do you think of the following UX: at install time Peergos asks the user which paths the app should have access to? 

The other useful property of this permission is that it covers the case where we don't understand the mimetype of some files.

**Comment by jonaswre** [comment]
Ahh okay I'm not to familiar with the sharing functions. Okay but that would work. 

I think only at install is not enough. There need to be a setting that can be changed. I think it would be best to present a setting in the launcher. Draw the folder tree with node expand/contract. Then allow to select the paths that are accessable.

**Comment by jonaswre** [comment]
And I think we might need this for write permissions sooner or later as well.

**Comment by ianopolous** [comment]
Ok cool. For an initial implementation we could just list the paths in the launcher, and let you add more or remove.

**Comment by kevodwyer** [comment]
folder picker completed: https://github.com/Peergos/web-ui/pull/489

**Comment by jonaswre** [comment]
I now not the right place but is there now a way to receive chats using websockets?

**Comment by ianopolous** [comment]
If you mean a realtime push api, rather than pull based, then we hope to add direct p2p streams where users can push arbitrary data over a browser to browser TLS connection. But it is still a ways off.

**Comment by jonaswre** [comment]
Exactly what I meant, so will be patient.
Btw Performance has gotten a lot better since I started using it. 

One annoying thing is that the PWA doesn't show Audio playback when in Background. I can press play and pause on my headphones but I can't get buttons in the notifications and if I open the app again I'm on the Homescreen not in the app I'm using for playback.

**Comment by ianopolous** [comment]
With any luck, the buttons issue should be fixed by this: https://github.com/Peergos/web-ui/pull/494
